### PR TITLE
Nice errors when deploying incompatible caas series

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

This PR needs https://github.com/juju/os/pull/2

When deploying a caas charm (or an iaas charm), we check the model tyype and error early and with a relevant message. We check both in the CLI (to error early) and in the backend to protect against general api calls.

## QA steps

Try to deploy a caas charm to an iaas model and visa versa.

### CODE REVIEW

File: cmd/juju/application/deploy_test.go [hunk 1]
```diff
@@ -495,6 +495,12 @@ func (s *DeploySuite) TestForceMachine(c *gc.C) {
 	s.assertForceMachine(c, machine.Id())
 }
 
+func (s *DeploySuite) TestInvalidSeriesForModel(c *gc.C) {
+	ch := testcharms.Repo.CharmArchivePath(s.CharmsPath, "dummy")
+	err := runDeploy(c, ch, "portlandia", "--series", "kubernetes")
+	c.Assert(err, gc.ErrorMatches, `series "kubernetes" in a non CAAS model not valid`)
```
Comment (howbazaar) [inline cmd/juju/application/deploy_test.go]: Don't say CAAS. I'd say "kubernetes"

File: core/model/model.go [hunk 2]
```diff
@@ -30,3 +36,24 @@ type Model struct {
...
+// ValidateSeries ensures the charm series is valid for the model type.
+func ValidateSeries(modelType ModelType, charmSeries string) error {
+	os, err := series.GetOSFromSeries(charmSeries)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	switch modelType {
+	case CAAS:
+		if !caasOS.Contains(os.String()) {
+			return errors.NotValidf("series %q in a CAAS model", charmSeries)
```
Comment (howbazaar) [inline core/model/model.go]: ditto

File: core/model/model.go [hunk 3]
```diff
@@ -30,3 +36,24 @@ type Model struct {
 	// ModelType is the type of model.
 	ModelType ModelType
 }
+
+var caasOS = set.NewStrings("Kubernetes")
```
Comment (howbazaar) [inline core/model/model.go]: Use the os constant.

File: cmd/juju/application/deploy_test.go (hunk 1 above)
Comment (wallyworld) [reply cmd/juju/application/deploy_test.go]: yeah, that occurred to me also. i thing "container" is better though

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**